### PR TITLE
SONAR-7003 Refactor batch issue tracking

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

**Comment by henryju** [comment]
LGTM

File: sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java [hunk 1]
```diff
@@ -98,10 +105,10 @@ private void relocateManualIssues(Input<RAW> rawInput, Input<BASE> baseInput, Tr
...
         }
         if (!Strings.isNullOrEmpty(baseHash)) {
-          int[] rawLines = rawInput.getLineHashSequence().getLinesForHash(baseHash);
-          if (rawLines.length == 1) {
-            tracking.keepManualIssueOpen(base, rawLines[0]);
-          } else if (rawLines.length == 0 && rawInput.getLineHashSequence().hasLine(base.getLine())) {
+          Set<Integer> rawLines = rawInput.getLineHashSequence().getLinesForHash(baseHash);
+          if (rawLines.size() == 1) {
+            tracking.keepManualIssueOpen(base, rawLines.iterator().next());
+          } else if (rawLines.size() == 0 && rawInput.getLineHashSequence().hasLine(base.getLine())) {
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java]: MAJOR Use isEmpty() to check whether the collection is empty or not. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1155)

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 2]
```diff
@@ -100,6 +108,29 @@ public void init() {
...
+
+    return new IssueTrackingInput<ServerIssueFromWs>(serverIssues, refHashes);
+  }
+
+  private static Input<TrackedIssue> createRawInput(Collection<TrackedIssue> rIssues, @Nullable SourceHashHolder sourceHashHolder) {
+    List<String> baseHashes;
+    if (sourceHashHolder != null && sourceHashHolder.getHashedSource() != null) {
+      baseHashes = Arrays.asList(sourceHashHolder.getHashedSource().hashes());
+    } else {
+      baseHashes = new ArrayList<String>(0);
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: MAJOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2293)

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 3]
```diff
@@ -100,6 +108,29 @@ public void init() {
...
+    }
+
+    return new IssueTrackingInput<TrackedIssue>(rIssues, baseHashes);
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: MAJOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2293)

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 4]
```diff
@@ -100,6 +108,29 @@ public void init() {
     return trackedIssues;
   }
 
+  private static Input<ServerIssueFromWs> createBaseInput(Collection<ServerIssueFromWs> serverIssues, @Nullable SourceHashHolder sourceHashHolder) {
+    List<String> refHashes;
+
+    if (sourceHashHolder != null && sourceHashHolder.getHashedReference() != null) {
+      refHashes = Arrays.asList(sourceHashHolder.getHashedReference().hashes());
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: BLOCKER NullPointerException might be thrown as 'getHashedReference' is nullable here [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 5]
```diff
@@ -100,6 +108,29 @@ public void init() {
...
+    } else {
+      refHashes = new ArrayList<String>(0);
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: MAJOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2293)

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 6]
```diff
@@ -100,6 +108,29 @@ public void init() {
...
+    }
+
+    return new IssueTrackingInput<ServerIssueFromWs>(serverIssues, refHashes);
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: MAJOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2293)

File: sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java [hunk 7]
```diff
@@ -149,42 +180,47 @@ private SourceHashHolder loadSourceHashes(BatchComponent component) {
...
+      if (unmatchedIssue.getRuleKey().isManual() && !Issue.STATUS_CLOSED.equals(unmatchedPreviousIssue.getStatus())) {
         relocateManualIssue(unmatched, unmatchedIssue, sourceHashHolder);
       }
       updateUnmatchedIssue(unmatched, false /* manual issues can be kept open */);
-      issues.add(unmatched);
+      mergeTo.add(unmatched);
+    }
+  }
+
+  private void addUnmatchedFromReport(Iterable<TrackedIssue> rawIssues, Collection<TrackedIssue> trackedIssues, Date analysisDate) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/issue/tracking/LocalIssueTracking.java]: MAJOR Make "addUnmatchedFromReport" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: sonar-core/src/main/java/org/sonar/core/util/UuidFactoryFast.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  private UuidFactoryFast() {
+    //
+  }
+
+  private static final char[] hexArray = "0123456789ABCDEF".toCharArray();
+  private static int sequenceNumber = 0;
+
+  @Override
+  public String create() {
+    int sequenceId = sequenceNumber++;
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidFactoryFast.java]: CRITICAL Make the enclosing method "static" or remove this set. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2696)
MAJOR Extract this increment or decrement operator into a dedicated statement. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* BLOCKER 1 blocker
* CRITICAL 1 critical
* MAJOR 7 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* BLOCKER 1 blocker
* CRITICAL 1 critical
* MAJOR 7 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* BLOCKER 1 blocker
* CRITICAL 1 critical
* MAJOR 7 major

Watch the comments in this conversation to review them.

File: sonar-core/src/main/java/org/sonar/core/util/UuidFactoryFast.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  public String create() {
+    long timestamp = System.currentTimeMillis();
+
+    byte[] uuidBytes = new byte[9];
+
+    // Only use lower 6 bytes of the timestamp (this will suffice beyond the year 10000):
+    putLong(uuidBytes, timestamp, 0, 6);
+
+    // Sequence number adds 3 bytes:
+    putLong(uuidBytes, sequenceNumber++, 6, 3);
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidFactoryFast.java]: MAJOR Extract this increment or decrement operator into a dedicated statement. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS881)
CRITICAL Make the enclosing method "static" or remove this set. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2696)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* BLOCKER 1 blocker
* CRITICAL 1 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: sonar-core/src/main/java/org/sonar/core/util/UuidFactoryFast.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    return byteArrayToHex(uuidBytes);
+  }
+
+  public static UuidFactoryFast getInstance() {
+    return instance;
+  }
+  
+  private static int getSequenceNumber() {
+    return sequenceNumber++;
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidFactoryFast.java]: CRITICAL Remove this increment or correct the code not to waste it. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2123)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* BLOCKER 1 blocker
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

**Pull request merged**